### ISSUE TITLE
Project: Typescript Excellency

### ISSUE THREAD

**Issue by Uzlopak** [Automattic/mongoose#11312]
This is a Thread about improving the mongoose typings significantly.

Currently typings are checked by very slow typescript transpilation and integration tests. To improve the dx of mongoose of use mongoose devs and of users of mongoose I plan to improve the typings in following steps:

- add tsd as typing validation -> #11311 
- move typings into types folder -> #11311 
- split the current big index.d.ts into more typing files (if possible) to handle the typings more easily
- improve typings and write corresponding tests in tsd to ensure that they are valid. Potentially check if documentation and typings are in sync (nice to have).
- make the typings more genetic and ensure that in fitting cases be respected by the other types. Goal is to only once pass the interface to the Schema and have everything correctly typed  even in aggregation pipeline and queries.
- maybe remove typescript transpilation and integration tests -> done

To improve the typings I plan:
- rename all callback-parameters, which are now either named cb or callback, from cb to callback
- sort the function overloads by decreasing arity. Currently alot of function overloads are in the wrong order decreasing the dx in vscode as typescript suggests from more specific to more general typing
- if an async function has a callback specified in the call, then don't type it with a returntype of Promise
- if an async function has no callback specified in the call, then type the returntype as Promise

https://www.typescriptlang.org/docs/handbook/declaration-files/do-s-and-don-ts.html

I think this will take about 2-4 weeks.

Looking for feedback.

**Comment by eroglu94** [comment]
Is this issue linked with Visual Studio Code slow intellisense(code completition) after versions for mongoose 5.11 installed? If so slow intellisense still a big problem in version 5.11 to 6.7.2 when they are installed.

**Comment by vkarpov15** [comment]
I think we can mark this particular task as done, we've already switched to tsd, split up `index.d.ts` into multiple files, etc.
